contributors/design-proposals/federated-statefulsets.md

```diff
@@ -0,0 +1,491 @@
+like listed further. The name resolution is exactly as it happens for the local statefulsets in
+k8s today. Additionally note that the federation name _myfederation_ has explicitly been dropped
+in this example to demonstrate a local query. This is also how the in cluster kube-dns works. It
+parses and considers the service name query to be local or federated global based on the presence
+of valid federation name segment in the queried string. [Refer here](https://github.com/kubernetes/dns/blob/master/pkg/dns/dns.go#L588).
+
+The CNAME query will return SRV records as below.
+
+within cluster c1
+```
+store-c1-0.etcd.mynamespace.svc.example.com
+store-c1-1.etcd.mynamespace.svc.example.com
+store-c1-2.etcd.mynamespace.svc.example.com
+store-c1-3.etcd.mynamespace.svc.example.com
+```
+within cluster c2
+```
+store-c2-0.etcd.mynamespace.svc.example.com
+store-c2-1.etcd.mynamespace.svc.example.com
+store-c2-2.etcd.mynamespace.svc.example.com
+store-c2-3.etcd.mynamespace.svc.example.com
+```
+within cluster c3
+```
+store-c3-0.etcd.mynamespace.svc.example.com
+store-c3-1.etcd.mynamespace.svc.example.com
+store-c3-2.etcd.mynamespace.svc.example.com
+```
+The SRV record will eventually map to an A record with a valid replica pod in-cluster IP if
+the replica is up and running.
+
+Additional pointer is that, if the given cluster is properly added to federation, this same
+headless service and related resolved names can be queried including the federation name in
+query too. For example the service CNAME record ```etcd.mynamespace.myfederation.svc.example.com```
+will also result in similar local names as elaborated above, especially because at this point
+the headless service, even though federated, will not get any dns names updated into the
+global (federated) dns server. The results will however be different, if the global identities
+are also requested by the user as elaborated in the next section.
+
+### Instance identity and discovery (global identities)
+
+The above is fine for the case of multiple local quora (one in each cluster), but these
+identities cannot work across the clusters, as the local pod IPs cannot be reached across
+the clusters.
+
+The proposal to make it work across the clusters is to create an additional service of type
+&#39;LoadBalancer&#39; for each stateful pod instance that is created in the clusters. The
+name of each service instance will be same as the stateful instance name assigned to the
+stateful replicas. This will be automatically created by the federated statefulset controller,
+on demand, proposed to be controlled by an annotation on the statefulset. If the annotation
+specifies so (users choice), the federation controller will create the additional service of
+type &#39;LoadBalancer&#39; against each pod replica. The key name for the annotation will be:
+```federation.kubernetes.io/statefulset-lb-needed```
+The values could be ```true``` or ```false```. The default value will be ```false```, considering
```

Comment:
I think a similar solution can be achieved using ingress or nodeports. A stable federated ingress solution out of the box seems not available across cloud providers. Nodeport based solution might not be portable across cloud provider specific dns servers. I might be missing somethings. Suggestions are welcome.